File: scenedescription.py
import math
import json, jsonpickle
from linearalgebra import *

# An axis-aligned boudning box
class BBOX():
    def __init__(self, center, extent):
        self.center = center
        self.extent = size / 2
        self.x_plane = Plane(Vector3(y=0.0, z=0.0), center.X)
        self.y_plane = Plane(Vector3(x=0.0, z=0.0), center.Y)
        self.z_plane = Plane(Vector3(x=0.0, y=0.0), center.Z)

    def contains(self, p_center, p_extent=Vector3() * 0):
        if p_center.x + p_extent.x <= self.center.x + self.extent.x and \
           p_center.x - p_extent.x >= self.center.x - self.extent.x and \
           p_center.y + p_extent.y <= self.center.y + self.extent.y and \
           p_center.y - p_extent.y >= self.center.y + self.extent.y and \
           p_center.z + p_extent.z <= self.center.z + self.extent.z and \
           p_center.z - p_extent.z >= self.center.z + self.extent.z:
           return True

class Material():
    def __init__(self,
                 diffuse,
                 diffuse_intensity,
                 specular,
                 specular_intensity,
                 ambient,
                 ambient_intensity,
                 gloss):
        self.d = diffuse
        self.d_int = diffuse_intensity
        self.s = specular
        self.s_int = specular_intensity
        self.a = ambient
        self.a_int = ambient_intensity
        self.g = gloss

    def shade(self, V, N, R, L, light):
        # diffuse
        d = light.intensity * light.color * self.d * (N ** L)
        # specular
        s = light.intensity * light.color * self.s * math.pow((V * R), self.g)
        # ambient
        a = self.a
        #combined
        return d * self.d_int + s * self.s_int + a * self.a_int

class Sphere():
    def __init__(self, pos=Vector3() * 0, radius=1.0):
        self.pos = pos
        self.radius = radius
        self.center = pos
        self.extent = (Vector3() * self.radius) / 2

    def intersect(self, ray):
        pass

    def get_normal(self, point):
        pass

class Cube():
    def __init__(self, pos=Vector3() * 0, size=1.0):
        self.pos = pos
        self.size = size
        self.center = pos
        self.extent = (Vector3() * self.size) / 2

    def intersect(self, ray):
        pass

class Plane():
    def __init__(self, normal=Vector3() * 0, distance=1.0):
        self.normal = normal
        self.distance = distance

    def intersect(self, ray):
        pass

class CustomOBJ():
    def __init__(self, pos=Vector3() * 0, filepath="test.obj"):
        self.transform = transform
        self.filepath = filepath

class Octree():
    def __init__(self, center=Vector3() * 0, size=1.0, objects=[]):
        children = []
        for object in objects:
            object_bbox = object.center

    def intersect(self, ray):
        pass

class Camera():
    def __init__(self,
                 pos=Vector3() * 0,
                 forward=Vector3(x=0.0, y=0.0, z=-1.0),
                 up=Vector3(x=0.0, z=0.0),
                 fov=40.0):
        self.pos = pos
        self.forward = forward
        self.up = up
        self.right = (forward % up).normalized()
        self.fov = fov
        self.transform = Matrix3(cols=[self.right,
                                         self.up,
                                         self.forward])

class PointLight():
    def __init__(self, pos=Vector3() * 0, color=Vector3(), intensity=1.0):
        self.pos = pos
        self.color = color
        self.intensity = intensity

    def intersect(self, origin):
        return False

class SceneDescription():
    def __init__(self, lights=[], objects=[]):
        self.lights = lights
        self.objects = objects

    def import_from_file(self, file):
        with open(file, 'r') as f:
            updated = jsonpickle.decode(json.load(f))
            self.__dict__.update(updated.__dict__)

    def export_to_file(self, file):
        with open(file, 'w') as f:
            f.write(json.dumps(jsonpickle.encode(self)))

    def flatten_to_octree(self):
        return Octree()

File: scenedescription.cpp
class Ray {
  public:
    Float3 origin, direction;
    float distance;
    Ray(Float3 o, Float3 dest) {
      origin = o;
      vector = (dest - o);
      direction = (vector).normalized();
      distance = (vector).magnitude();
    }
    Ray(Float3 o, Float3 dir, Float3 dist):
      origin(o), direction(dir), distance(dist) {}
};

class Hit {
  public:
    Float3 pos, V, N, R;
    Hit(): Hit(Float3(), Float3(), Float3()) {}
    Hit(Float3 p, Float3 v, Float3 n): pos(p), V(v), N(n) {
      R = (2 * (V^N))*N - V;
    }
    float vector_to(Float3 point) {
      return (pos - point).normalized();
    }
};

class Plane {
  public:
    Float3 normal;
    float distance;
    Plane(Float3 norm, float dist): normal(norm), distance(dist) {}
    Hit intersect(Ray ray) {
      return Hit();
    }
};

class BBOX {
  public:
    Float3 center, extent;
    Plane x_plane, y_plane, z_plane;
    BBOX(Float3 ctr, Float3 ext) {
      center = ctr;
      extent = ext;
      x_plane = Plane(Float3(0.0, 0.0, center.x));
      y_plane = Plane(Float3(0.0, 0.0, center.y));
      z_plane = Plane(Float3(0.0, 0.0, center.z));
    }
    bool contains(Float3 p_center, Float3 p_extent) {
      if (p_center.x + p_extent.x <= center.x + extent.x &&
          p_center.x - p_extent.x >= center.x - extent.x &&
          p_center.y + p_extent.y <= center.y + extent.y &&
          p_center.y - p_extent.y >= center.y - extent.y &&
          p_center.z + p_extent.z <= center.z + extent.z &&
          p_center.z - p_extent.z >= center.z - extent.z &&) {
        return true;
      }
    }
    bool intersect(Ray ray) {
      return false;
    }
};

class Material {
  public:
    Float3 d, s, a;
    float di, si, ai, g;
    Material(Float3 D, Float3 S, Float3 A,
             float Di, float Si, float Ai, float G):
             d(D), s(S), a(A), di(Di), si(Si), g(G) {}
    // li = light intensity, lc = light color
    Float3 shade(Hit h, Light light) {
      Float3 diffuse = (light.color * light.intensity) * (d * (h.N ^ h.vector_to(light.pos)));
      Float3 specular = (light.color * light.intensity) * (s * pow(h.V * h.R), g));
      Float3 ambient = a;
      return (diffuse * di) + (specular * di) + (ambient * ai);
    }
};

class SceneObject {
  protected:
    BBOX bound;
  public:
    BBOX get_bound() {
      return bound;
    }
};

class Sphere : public SceneObject {
  protected:
    Float3 center;
    float radius;

  public:
    Sphere(Float3 c, float r) {
      center = c;
      radius = r;
      bound = Float3(radius);
    }
    void set_center(Float3 c) {
      center = c;
    }

    Float3 get_center() {
      return center;
    }

    void set_radius(float radius) {
      radius = r;
      bound = Float3(radius);
    }

    float get_radius() {
      return radius;
    }

    Hit intersect(Ray ray) {
      return Hit();
    }
};
